docs/cs-proof.md

```diff
@@ -0,0 +1,800 @@
+
+There are two kinds of variables in the constraint system:
+
+1. \\(m\\) **high-level** variables, the input secrets \\(\mathbf{v}\\),
+2. \\(n\\) **low-level** variables, the internal inputs and outputs of the [multiplication gates](#multiplication-gates): \\(\mathbf{a}\_L, \mathbf{a}\_R, \mathbf{a}\_O\\).
+
+### Multiplication gates
+
+Each multiplication gate takes two input variables and multiplies them to get an output.
+That relation for all multiplication gates is represented by \\(n\\) constraints:
+
+\\[
+\mathbf{a}\_L \circ \mathbf{a}\_R = \mathbf{a}\_O,
+\\]
+where:
+* \\(\mathbf{a}\_L\\) is the vector of the first input to each gate,
+* \\(\mathbf{a}\_R\\) is the vector of the second input to each gate,
+* \\(\mathbf{a}\_O\\) is the vector of multiplication results.
+
+### Linear constraints
+
+Linear constraints are expressed using a vector of \\(q\\) equations that use linear combinations of the [variables](#variables):
+
+\\[
+\mathbf{W}\_L \cdot \mathbf{a}\_L +
+\mathbf{W}\_R \cdot \mathbf{a}\_R +
+\mathbf{W}\_O \cdot \mathbf{a}\_O =
+\mathbf{W}\_V \cdot \mathbf{v} +
+\mathbf{c},
+\\]
+where:
+
+* \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O\\) are weights applied to the respective inputs and outputs of the [multiplication gates](#multiplication-gates) (low-level [variables](#variables)),
+* \\(\mathbf{W}\_V\\) are weights applied to the high-level [variables](#variables) \\(\mathbf{v}\\),
+* \\(\mathbf{c}\\) is a vector of constant terms used in the linear constraints.
+
+
+### Building constraints
+
+Bulletproofs framework allows building constraint systems _on the fly_, without a trusted setup.
+This allows instantiating constraints from a _family of constraint systems_ parametrized by
+public values and commitments to the secret inputs.
+As a result, the instantiation can be thought of as a _challenge_ generated by a verifier to a prover.
+
+The prover starts out by committing to its secret inputs \\(\mathbf{v}\\)
+and obtaining \\(m\\) _variables_ representing these inputs.
+
+Then, the prover performs a combination of the following operations to generate the constraints:
+
+1. **Allocate a multiplier:** a new [multiplication gate](#multiplication-gates) is added, represented by three [low-level variables](#variables) \\(a_L, a_R, a_O\\), for left input, right input and output value respectively.
+2. **Add a constraint:** a [linear combination](#linear-constraints) of any number of [variables](#variables) is encoded into appropriate positions in matrices \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\) and a vector of constants \\(\mathbf{c}\\).
+3. **Request a challenge scalar:** a random scalar returned in response to committed [high-level variables](#variables).
+
+
+### Uncommitted variables
+
+Often a constraint system is composed of _gadgets_: subsets of constraints
+that implement features of a higher-level protocol.
+The “wires” that connect the gadgets are called _uncommitted variables_.
```

Comment:
Also, I think mentioning gadgets here (before explaining what a gadget is) is confusing... maybe better to not talk about gadgets in this section

Reply:
WDYT?

```
### Gadgets

Gadgets are buildings blocks of a constraint system that map to some functions in a higher-level protocol.
For example, a **shuffle gadget** creates constraints that prove that two sets of variables are equal up to a permutation. Another example is a **range proof gadget** that checks that a value is composed of a specific number of bits.

Gadgets receive some [variables](#variables) as inputs, may [allocate more variables](#building-constraints) for internal use,
and produce constrains involving all these variables.


### Uncommitted variables

Often a [gadget](#gadgets) needs an internal variable to connect with another gadget,
or to implement its internal logic, without requiring a distinct [high-level variable](#variables) commitment \\(V\_i\\) for it.
Such **uncommitted variables** are created from left and right variables \\(a\_L, a\_R\\) of additional multiplication gates.
Output variables \\(a\_O\\) are not used for this purpose because
they are implicitly constrained by a [multiplication gate](#multiplication-gates)
and cannot be used as independent uncommitted variables.

Note: uncommitted variables have their name due to lack of the individual commitments,
but they are still committed collectively with all [low-level variables](#variables)
using a single vector Pedersen commitment \\(A\_I\\) as required by the underlying proof protocol.
```

Reply:
that looks good!

Reply:
reworded once more:

<img width="1176" alt="image" src="https://user-images.githubusercontent.com/698/46771964-af28e680-ccaa-11e8-8c9c-088853dd5976.png">